app/services/db/private-cloud-costs.ts

```diff
@@ -0,0 +1,164 @@
+import { DecisionStatus } from '@prisma/client';
+import { namespaceKeys } from '@/constants';
+import prisma from '@/core/prisma';
+
+export interface CostItem {
+  startDate: Date;
+  endDate: Date;
+  cpu: number;
+  storage: number;
+  cpuCost: number;
+  storageCost: number;
+  totalCost: number;
+  unitPriceId?: string;
+}
+
+export async function getUnitsPrice() {
+  const unitPrice = await prisma.privateCloudUnitPrice.findMany({
+    orderBy: { date: 'desc' },
+  });
+  return unitPrice;
+}
+
+export async function getCostItemsForRange(licencePlate: string, startDate: Date, endDate: Date) {
+  const unitPrices = await getUnitsPrice();
+
+  const allRequests = await prisma.privateCloudRequest.findMany({
+    where: {
+      licencePlate,
+      isQuotaChanged: true,
+      decisionStatus: { in: [DecisionStatus.PROVISIONED] },
+      provisionedDate: { not: null, lte: endDate },
+    },
+    include: { decisionData: true },
+    orderBy: { provisionedDate: 'asc' },
+  });
+
+  const requestsInRange = allRequests.filter(
+    (req) => req.provisionedDate! >= startDate && req.provisionedDate! <= endDate,
+  );
+
+  if (requestsInRange.length === 1) {
+    const fallback = [...allRequests].reverse().find((req) => req.provisionedDate! < startDate);
+    if (fallback) return [fallback, ...requestsInRange] as unknown as CostItem[];
+  }
+
+  const costItems: CostItem[] = [];
+
+  for (const r of requestsInRange) {
+    const provisionedDate = r.provisionedDate!;
+    let applicablePrice = [...unitPrices]
+      .sort((a, b) => new Date(b.date).getTime() - new Date(a.date).getTime())
+      .find((price) => new Date(price.date) <= provisionedDate);
+
+    if (!applicablePrice) {
+      applicablePrice = unitPrices[unitPrices.length - 1];
```

Comment:
if there is no unit price added before the request provisioned then, the unit price should be `0`